### PR TITLE
Refactor `brew bump`

### PR DESCRIPTION
**Pull request by nandahkrishna**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This PR refactors `brew bump` to accomplish the following:

* Display packages iteratively rather than all at once. Note: Earlier, if a package was present only in Homebrew's repositories on Repology, `bump` would quietly ignore it irrespective of `livecheck`'s output. Now, a message is displayed stating the same (eg: `airshare`, Note: Due to an earlier issue with PyPI repology/repology-updater#278).

* Use methods from `Livecheck` rather than naively parsing `brew livecheck`'s output.

* Refactor `utils/repology` and remove `utils/livecheck_formula`.

* Minor changes to Livecheck.

Concerns:

* Handling exceptions/failures in `bump` in case Repology API or livecheck error out.

* Do we need additional debug/verbose/json output?

To Do:

- [x] Some comments (marked as TODO).

- [x] Fix issues where the package name on Repology is different from the Formula name (eg: `acme`, Note: the Repology-reported version is incorrect too) -- requested Repology and waiting for the `project-by` search tool to be enabled for Homebrew.

Do let me know of any suggestions, or issues with this implementation. Thanks!

CC @samford @SeekingMeaning

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by samford** [comment]
I've been working on moving the livecheck docs PR forward (I finally pushed the latest draft) but I'll take a look at this when I have a moment.

> Do we need additional debug/verbose/json output?

A few months ago, I spent some time working on JSON output for `brew bump` but I never created a PR for it. I was holding on to it because I had plans to eventually refactor `brew bump` like you've done here (though I'm certainly glad you beat me to it).

Having good JSON output is very useful, so I'll have to revisit my previous work when I get a chance. I believe it was pretty much finished when I last left it and I would just need to rebase it onto this.

**Comment by nandahkrishna** [comment]
Repology now supports Homebrew support for their `project-by` API, so URLs of the form `https://repology.org/tools/project-by?repo=homebrew&name_type=srcname&target_page=api_v1_project&name=<NAME>` can now be used to get package data by Formulae names, rather than their Repology names.

This helps us handle issues with Formulae such as `acme`, which is listed as `acme-crossassembler` on Repology. Earlier, `brew bump` would not return any output for this when using `brew bump acme`, but would output information when using just `brew bump` if `acme` was outdated.

I'll be adding a few changes here soon to address this.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by nandahkrishna** [comment]
Just waiting on #10208 (changes to made here as a result) and Sam's review before going ahead and merging this.

**Comment by samford** [comment]
I merged #10208 about an hour ago, so you're free to move forward there. I'll work on reviewing this in the interim time.

**Review by samford — DISMISSED** [review]
There's still some room for improvement in `brew bump` but this resolves my primary concern (calling `brew livecheck` instead of using livecheck methods). I've left suggestions/comments for a few smaller areas.

Another notable area where `brew bump` can be improved beyond this is to only fetch pull request data once per repository, instead of fetching it for every formula. When running bump without an API token, I easily hit the GitHub API rate limit after only ~10 formulae (giving `Open pull requests: none` at that point). This is something we can address in a follow-up PR but I wanted to mention it, at least.

File: Library/Homebrew/dev-cmd/bump.rb [hunk 1]
```diff
@@ -29,64 +29,107 @@ def bump_args
...
     end
 
-    display(validated_formulae)
+    version_info = Livecheck.latest_version(
+      formula,
+      json: true, full_name: false, verbose: false, debug: false,
+    )
+    latest = version_info[:latest] if version_info.present?
+
+    return "not found" if latest.blank?
```
[the remark this answers is not in the document]
Comment (nandahkrishna) [reply Library/Homebrew/dev-cmd/bump.rb:101]: I think `unable to get versions` makes more sense to someone using `brew bump`. I've made it lowercase for consistency.

File: Library/Homebrew/livecheck/livecheck.rb [hunk 2]
```diff
@@ -60,6 +60,30 @@ def livecheck_strategy_names
       @livecheck_strategy_names.freeze
     end
 
+    # Identify taps other than homebrew/core and homebrew/cask in use
+    # for current formulae and casks and load additional Strategy
+    # files them.
```
[the remark this answers is not in the document]
Comment (nandahkrishna) [reply Library/Homebrew/livecheck/livecheck.rb:65]: This is more clear, thanks!

File: Library/Homebrew/livecheck/livecheck.rb [hunk 3]
```diff
@@ -60,6 +60,30 @@ def livecheck_strategy_names
...
+    sig do
+      params(formulae_and_casks_to_check: T::Enumerable[T.any(Formula, Cask::Cask)]).void
+    end
```
[the remark this answers is not in the document]
Comment (nandahkrishna) [reply Library/Homebrew/livecheck/livecheck.rb:68]: Strangely I didn't get this suggestion. I'll change it though.

**Comment by nandahkrishna** [comment]
Thanks for reviewing, Sam. I recently noticed the rate limiting issue too so we should address it in a follow-up PR. Let me know if you have any other suggestions, or if this is good to merge.

**Comment by samford** [comment]
> Let me know if you have any other suggestions, or if this is good to merge.

I don't have any other specific suggestions and I'm fine with this being merged when it passes CI. It's currently failing CI due to the code coverage, so it may be a good idea to try to increase coverage where we can by reworking/expanding the tests.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Great work here, thanks again @nandahkrishna!

**Pull request merged**